Prevent substitution inside Derivative in some cases

normalhuman opened the pull request:
#### References to other Issues or PRs

Fixes #11737. Fixes #13791. Fixes #13795. Fixes #13848.

#### Brief description of what is fixed or changed

Currently, a substitution such as `Derivative(f(x, y), y).subs(y, x)` results in `Derivative(f(x, x), x)` which is a different derivative. This affects the computation of derivatives of orders 2 or greater of expressions with undefined functions. This patch modifies _eval_subs method of Derivative class to disallow substitution when it is clearly wrong to do.

This is an alternative approach to #13801 which attempted to codify all _allowed_ substitutions; this one codifies _disallowed_ ones as follows. Substitution (old, new) cannot be done inside Derivative(expr, vars) if it is:

1. changing expr by introducing a variable among vars, or
2. changing vars by introducing a variable contained in expr

Exception: complete replacement of expr by new (expr == old) is always allowed, so for example `Derivative(f(x), x).subs(f(x), y).subs(y, f(x))` results in Derivative(f(x), x) although one may call this abuse of notation. This is something that `solve` relies on:

> when an object other than a Symbol is given as a symbol, it is isolated algebraically and an implicit solution may be obtained. This is mostly provided as a convenience to save one from replacing the object with a Symbol and solving for that Symbol. It will only work if the specified object can be replaced with a Symbol using the subs method.
```
>> solve(f(x).diff(x) - f(x) - x, f(x))
[-x + Derivative(f(x), x)]
```
Temporarily replacing an expression being differentiated by a dummy symbols seems a reasonable thing to do. 

The rules 1 and 2 still allow for mathematically questionable but convenient operations of replacing a function being differentiated. Some such operations are baked into Physics module, e.g., 
```
Derivative((-R*u3(t) - b*u1(t) + u2(t))/R, t).subs(u3(t), 0)
Derivative(r*Derivative(f(r, th), r), r).subs(f(r, th), r**2*sin(th))  
```
where the authors expect the substitution to be carried out. The logic is that first a general expression with derivatives is formed, and then a specific function is put in to replace an abstract one.

Another example, found many times in `test_ode`, is applying `subs(f(x), sol)` to a differential equation such as `Eq(diff(f(x), x), 2*f(x))`. Technically, this substitution does not commute with the derivative here, but the meaning is clear and the substitution is allowed to proceed, so that Eq evaluates to True.

normalhuman committed:
Prevent substitution inside Derivative in some cases

Currently, a substitution such as Derivative(f(x, y), y).subs(y, x) results in
Derivative(f(x, x), x) which is a different derivative. This affects the computation
of derivatives of orders 2 or greater of expressions with undefined functions.
This commit modifies _eval_subs method of Derivative class to disallow substitution
when it is clearly wrong. Fixes #13791; also fixes #13795.

Substitution (old, new) cannot be done inside Derivative(expr, vars) if it is:
1. changing expr by introducing a variable among vars, or
2. changing vars by introducing a variable contained in expr
Exception: complete replacement of expr by new (expr == old) is always allowed,
so for example Derivative(f(x), x).subs(f(x), y).subs(y, f(x)) results in
Derivative(f(x), x) although one may call this abuse of notation.

normalhuman committed:
Add another test for the correctness of diff on undefined functions

`f(x, f(x, x)).diff(x, 2)` is an example in which diff currently
gives a wrong answer. The answer is checked by replacing f with an
explicit function.

normalhuman commented:
I'm okay with removing the full expression exception. Only one test fails, [this test](https://github.com/sympy/sympy/blob/master/sympy/core/tests/test_subs.py#L490):
```
 assert Derivative(f(x), x).subs(f(x), y).subs(y, f(x)) == Derivative(f(x), x)
```
The first subs results in `Derivative(y, x)` and then the second subs tries to change `y` back to `f(x)` but that is not allowed because it's introducing x. Should we consider the above test as abuse of subs and remove the exception that allows it?

gxyd reviewed (commented):
nitpicks

normalhuman committed:
Cleaner syntax, more consistent with SymPy norms

normalhuman committed:
Replace Subs with subs in Derivative.__new__

Now that subs is sanity-checked, the Derivative.__new__ method
should not shy away from using it (the shyness does not help in
avoiding errors anyway, because doit() will execute the substitution)

asmeurer commented:
Found an old issue that was related to this https://github.com/sympy/sympy/issues/4703. So it turns out this has always been a problem, but it was fixed in the past by making Derivative not evaluate with subs (I remember that behavior being needed in the ode module; I don't know if it is used any more, though).

normalhuman commented:
I can probably get rid of the complete-replacement exception by changing [`recast_to_symbols`](https://github.com/sympy/sympy/blob/master/sympy/solvers/solvers.py#L61) method of `solve` to use xreplace instead of subs. Would that be preferable?

normalhuman committed:
Remove the complete-expression exception for subs

It was previously used by solve so that when solving, e.g.,
f(x) + Derivative(f(x), x) = 3
it could temporarily replace f(x) by a dummy symbol, and
eventually put it back, presenting the solution as
Eq(f(x), 3 - Derivative(f(x), x))
This, specifically the restoration step, will now require xreplace.

asmeurer commented:
I'm not completely sure. Requiring xreplace for these kinds of things is sounding better and better, though.

normalhuman commented:
Well, I got rid of the exception, put `xreplace` where it was needed, and nothing exploded. It'd be nice to get more input on the issue, I obviously think this is a good way of fixing `_eval_subs` but I'm biased.
